pkg/executor/mpperr/mpp_err_recovery.go

```diff
@@ -0,0 +1,201 @@
+	return m.holder.capacity > 0 && !m.holder.alreadyFulled
+}
+
+// HoldResult tries to hold mpp result. You should call Enabled() and CanHoldResult() to check first.
+func (m *RecoveryHandler) HoldResult(chk *chunk.Chunk) {
+	m.holder.insert(chk)
+}
+
+// NumHoldChk returns the number of chunk holded.
+func (m *RecoveryHandler) NumHoldChk() int {
+	return len(m.holder.chks)
+}
+
+// NumHoldRows returns the number of chunk holded.
+func (m *RecoveryHandler) NumHoldRows() uint64 {
+	return m.holder.curRows
+}
+
+// PopFrontChk pop one chunk.
+func (m *RecoveryHandler) PopFrontChk() *chunk.Chunk {
+	if !m.enable || len(m.holder.chks) == 0 {
+		return nil
+	}
+	chk := m.holder.chks[0]
+	m.holder.chks = m.holder.chks[1:]
+	m.holder.memTracker.Consume(-chk.MemoryUsage())
+	return chk
+}
+
+// ResetHolder reset the dynamic data, like chk and recovery cnt.
+// Will not touch other metadata, like enable.
+func (m *RecoveryHandler) ResetHolder() {
+	m.holder.reset()
+}
+
+// RecoveryCnt returns the recovery count.
+func (m *RecoveryHandler) RecoveryCnt() uint32 {
+	return m.curRecoveryCnt
+}
+
+// Recovery tries to recovery error. Reasons that cannot recovery:
+//  1. Already return result to client because holder is full.
+//  2. Recovery method of this kind of error not implemented or error is not recoveryable.
+//  3. Retry time exceeds maxRecoveryCnt.
+func (m *RecoveryHandler) Recovery(info *RecoveryInfo) error {
+	if !m.enable {
+		return errors.New("mpp err recovery is not enabled")
+	}
+
+	if info == nil || info.MPPErr == nil {
+		return errors.New("RecoveryInfo is nil or mppErr is nil")
+	}
+
+	if m.holder.alreadyFulled {
+		return errors.New("mpp result holder is full")
+	}
+
+	if m.curRecoveryCnt >= m.maxRecoveryCnt {
+		return errors.Errorf("exceeds max recovery cnt: cur: %v, max: %v", m.curRecoveryCnt, m.maxRecoveryCnt)
+	}
+
+	m.curRecoveryCnt++
+
+	for _, h := range m.handlers {
+		if h.chooseHandlerImpl(info.MPPErr) {
+			return h.doRecovery(info)
+		}
+	}
+	return errors.New("no handler to recovery this type of mpp err")
+}
+
+type handlerImpl interface {
+	chooseHandlerImpl(mppErr error) bool
+	doRecovery(info *RecoveryInfo) error
+}
+
+var _ handlerImpl = &memLimitHandlerImpl{}
+
+type memLimitHandlerImpl struct {
+	useAutoScaler bool
+}
+
+func newMemLimitHandlerImpl(useAutoScaler bool) *memLimitHandlerImpl {
+	return &memLimitHandlerImpl{
+		useAutoScaler: useAutoScaler,
+	}
+}
+
+func (h *memLimitHandlerImpl) chooseHandlerImpl(mppErr error) bool {
+	if strings.Contains(mppErr.Error(), memLimitErrPattern) && h.useAutoScaler {
+		return true
+	}
+	return false
+}
+
+func (*memLimitHandlerImpl) doRecovery(info *RecoveryInfo) error {
+	// Ignore fetched topo, because AutoScaler will keep the topo for a while.
+	// And the new topo will be fetched when dispatch mpp task again.
+	if _, err := tiflashcompute.GetGlobalTopoFetcher().RecoveryAndGetTopo(tiflashcompute.RecoveryTypeMemLimit, info.NodeCnt); err != nil {
```

Comment:
The new topo is ready after this rpc, or do we need to do some backoff to make sure the new topo is ready during retry

Reply:
RecoveryAndGetTopo() will block until new topo is ready. And if timeout, will report error